core/src/main/java/org/elasticsearch/index/engine/Segment.java

```diff
@@ -167,12 +187,106 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalBoolean(compound);
         out.writeOptionalString(mergeId);
         out.writeLong(memoryInBytes);
-        
+
         boolean verbose = ramTree != null;
         out.writeBoolean(verbose);
         if (verbose) {
             writeRamTree(out, ramTree);
         }
+        if (out.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
+            writeSegmentSort(out, segmentSort);
+        }
+    }
+
+    Sort readSegmentSort(StreamInput in) throws IOException {
+        int size = in.readVInt();
+        if (size == 0) {
+            return null;
+        }
+        SortField[] fields = new SortField[size];
+        for (int i = 0; i < size; i++) {
+            String field = in.readString();
+            byte type = in.readByte();
+            if (type == 0) {
+                Boolean missingFirst = in.readOptionalBoolean();
+                boolean max = in.readBoolean();
+                boolean reverse = in.readBoolean();
+                fields[i] = new SortedSetSortField(field, reverse,
+                    max ? SortedSetSelector.Type.MAX : SortedSetSelector.Type.MIN);
+                if (missingFirst != null) {
+                    fields[i].setMissingValue(missingFirst ?
+                        SortedSetSortField.STRING_FIRST : SortedSetSortField.STRING_LAST);
+                }
+            } else {
+                Object missing = in.readGenericValue();
```

Comment:
I think it's worth as it would be more simple to understand code wise. I'll do that